### DISCUSSION TITLE
Bech32 length limitations

### DISCUSSION THREAD

**Discussion by msparks** [FiloSottile/age#452] [Spec feedback]
Hey, I ran into this limitation while working on an age plugin. Currently, age rejects plugin identities that are 'too long', i.e., more than 43 bytes of data. That's *technically correct* per the [draft plugin specification](https://github.com/C2SP/C2SP/pull/5/commits/e3f35d6d60fa9bda361f6f5487e8b70e7c2631b3?short_path=07bf8cc#diff-07bf8cc6707dbc77590da5e60a0e91a84362d6c7d5079af1fd7dbc2bb2cd2f14), but in practice it creates complexity for plugins, so it seems worth a discussion, especially since it may mean changes to the spec.

### Background

The Bech32 encoding, as defined in [BIP-0173](https://github.com/bitcoin/bips/blob/master/bip-0173.mediawiki#Bech32), produces strings that are **at most 90 characters long**, composed of 1–83 characters for HRP, 1 character for the separator ("1"), and at least 6 characters for the data and checksum.

The draft age plugin spec [requires Bech32 encoding for recipients and identities](https://github.com/C2SP/C2SP/pull/5/commits/e3f35d6d60fa9bda361f6f5487e8b70e7c2631b3?short_path=07bf8cc#diff-07bf8cc6707dbc77590da5e60a0e91a84362d6c7d5079af1fd7dbc2bb2cd2f14):

> -   Plugin-compatible recipients are encoded using Bech32 with the HRP `age1name` (lowercase).
> -   Plugin-compatible identities are encoded using Bech32 with the HRP `AGE-PLUGIN-NAME-` (uppercase).

And although Bech32 implementations vary in terms of HRP and data length enforcement, a strict reading and implementation of the age plugin specification would indeed be extremely limiting: age plugin identities can have *only 43 bytes of data* in the best case.

To illustrate:
- HRP: `AGE-PLUGIN-X-`: 13 characters (at least)
- Separator: `1`: 1 character
- Checksum: 6 characters
- Data: (90 - 13 - 1 - 6) * (5 bits / 8 bits for encoding overhead) = *43 bytes* for raw data.

### Problem statement

The Bech32 encoding is too restrictive for plugin recipients and identities. While some plugins can be implemented within the limits, such as [`age-plugin-yubikey`](https://github.com/str4d/age-plugin-yubikey), others using longer keys and/or storing metadata in identities cannot.

My use case is a `wrap` plugin that encrypts and decrypts a native age X25519 identity transparently, using age itself and potentially other plugins. The goal is to get an X25519 recipient, which can be used without plugins, but with an identity that is protected by a hardware device, for example. Thus, plugins are required for decryption, but not encryption. How the identity is wrapped is arbitrary and can include interesting operations like [splitting the identity](https://en.wikipedia.org/wiki/Shamir%27s_Secret_Sharing) or adding a [timelock](https://github.com/drand/tlock). It's possible to do this now with multiple `age` invocations (decrypt the identity first, then decrypt the message with the decrypted identity), but a plugin would make it transparent and therefore a better user experience. Specifically, the recipient is native X25519 `age1...` and the identity is `AGE-PLUGIN-WRAP-1...`. The problem is that a wrapped identity easily can be hundreds of bytes, so implementing this plugin is currently infeasible.

Other plugins will likely run into the same problem. One example is Kʏʙᴇʀ, discussed in https://github.com/FiloSottile/age/discussions/231. @FiloSottile experimented with a Kʏʙᴇʀ768+X25519 plugin and tweeted an example recipient, which is *very long*: https://twitter.com/FiloSottile/status/1544803635237998592.

### Existing implementations

As mentioned above, implementations vary:
- `age` uses a slightly modified version of the [reference Go implementation](https://github.com/sipa/bech32/blob/7a7d7ab158db7078a333384e0e918c90dbc42917/ref/go/src/bech32/bech32.go) of the Bech32 specification. Both implementations are strict and enforce the 90-character limit on [encode](https://github.com/FiloSottile/age/blob/8328d19d3e79846a99d432799b1ae31deef54c12/internal/bech32/bech32.go#L114-L116) and [decode](https://github.com/FiloSottile/age/blob/8328d19d3e79846a99d432799b1ae31deef54c12/internal/bech32/bech32.go#L147-L149).
- `rage` [uses](https://github.com/str4d/rage/blob/f2507197ca962850f37a5e82e5c3c6f6ca523e1b/age/Cargo.toml#L43-L44) the [`bech32` crate](https://docs.rs/bech32/latest/bech32/). It enforces the length on [the HRP](https://docs.rs/bech32/0.9.1/src/bech32/lib.rs.html#366) but not [the data](https://docs.rs/bech32/0.9.1/src/bech32/lib.rs.html#405).
  - Note: The `bech32` crate is not the reference implementation. Interestingly, the reference implementation in Rust has a length limitation [only on decode](https://github.com/sipa/bech32/blob/7a7d7ab158db7078a333384e0e918c90dbc42917/ref/rust/src/bech32.rs#L98-L100). The same is true for the [reference Python implementation](https://github.com/sipa/bech32/blob/7a7d7ab158db7078a333384e0e918c90dbc42917/ref/python/segwit_addr.py#L80-L81) and others (but not Go!).

### Potential solutions

The simplest approach here is to specify the maximum acceptable length of Bech32-encoded data, or remove the limitation explicitly, in the age plugin spec. The length limitation in Bech32 exists to make guarantees on the error detection of the checksum (see the end of the [*Checksum design* section in BIP-0173](https://github.com/bitcoin/bips/blob/master/bip-0173.mediawiki#Bech32)), but age identities and recipients perhaps have weaker requirements for the checksum strength than the motivating use case for BIP-0173, so allowing longer data lengths is reasonable.

Note that this approach would mean that the encoding isn't *strict* Bech32, so plugin authors would be unable to use strict Bech32 libraries like the reference Go implementation. That'll be confusing, but likely minor in practice.

Some possible alternatives:

- Remove the length limitation for identities only, with the assumption that identities are less prone to errors than recipients. This would be insufficient for the Kʏʙᴇʀ768+X25519 use case.
- Use side channels in plugins (e.g., environment variables, files) for longer identities, and keep only a stub or index in the identity file that `age` reads. Mentioned only for completeness; this option goes against age's simplicity philosophy.
- Support a different encoding scheme for identities. Mentioned for completeness.

**Comment by FiloSottile** [comment]
Thank you for the detailed investigation, and for addressing the potential issues @msparks, I really appreciate it!

I think I agree, such long recipients/identities are going to be copy-pasted, so I am less concerned about the error-correction effectiveness. I'll drop the length restriction before v1.1.0 final. @str4d, do you agree? You might want to add a test to lock this in.

**Comment by str4d** [comment]
Yes, I'm fine with this. Bech32 has been used almost since its inception with a longer data length limit: the Lightning Network raised the limit to 1024 bytes (the next cliff in Bech32) to accommodate their invoice data structures, and this limit (or lack thereof) has propagated elsewhere. I expect 1024 bytes to likely be sufficient here, but I also wouldn't object to just removing the limit.

**Comment by michaelforney** [comment]
Perhaps it should be mentioned in the specification somewhere that age uses a relaxed version of Bech32 without length limitations? I had the same concerns as @msparks, since there isn't any mention of the length restriction in https://c2sp.org/age or https://c2sp.org/age-plugin.

As is, it states that recipients and identities are encoded with Bech32 with a link to BIP-0173, which defines
> A Bech32 string is at most 90 characters long.

It was only after some digging that I found this discussion and learned that larger recipients and identities are acceptable.

My worry is for potential incompatibilities between clients written to the spec and plugins requiring large strings, or that potential plugin authors might deem their scheme incompatible with the spec and be deterred.

**Reply by BenWestgate** [reply]
long codex32 in BIP-0093 has a checksum that can support 1023 bech32 characters detecting 8 errors and correcting 4.

**Reply by FiloSottile** [reply]
The spec now says

> Bech32 is as specified in [BIP173](https://github.com/bitcoin/bips/blob/master/bip-0173.mediawiki), but without length limits on the data part.

**Reply by BenWestgate** [reply]
Sounds good. lightning invoices did this as well.
The reason it was limited to 90 characters was to guarantee detecting 4 errors.
> Even though the chosen code performs reasonably well up to 1023 characters, other designs are preferable for lengths above 89 characters (excluding the separator).

**Reply by mlemanowicz** [reply]
> The spec now says
> 
> > Bech32 is as specified in [BIP173](https://github.com/bitcoin/bips/blob/master/bip-0173.mediawiki), but without length limits on the data part.

I have two comments:

1. The the age implementation seems to permit over-the-limit HRPs, as can be seen in the following test case:
    https://github.com/FiloSottile/age/blob/main/internal/bech32/bech32_test.go#L50
    This violates the Bech32 spec which defines HRP length within the [1, 83] interval.
    The age spec should include that deviation or else a length check of the non-data part should be added to the implementation.
    Relevant commit: https://github.com/FiloSottile/age/commit/7354aa0d08a06eac42c635670a55f858bd23c943

2. The age implementation of Encode can emit uppercase Bech32 strings:
    https://github.com/FiloSottile/age/blob/main/internal/bech32/bech32.go#L107
    This also violates the Bech32 spec which states
    >Encoders MUST always output an all lowercase Bech32 string.
    
    It might be sensible to document that too.

**Reply by FiloSottile** [reply]
> The age spec should include that deviation

The text you quoted is exactly me saying we added "but without length limits on the data part" to the spec.

The same paragraph of the spec also says "Note that Bech32 strings can only be all uppercase or all lowercase, but the checksum is always computed over the lowercase string."

**Reply by mlemanowicz** [reply]
> "but without length limits on the data part"

According to [BIP173](https://github.com/bitcoin/bips/blob/master/bip-0173.mediawiki#bech32), the *data part* and the *human-readable part* are separate sections.
The data part only refers to the payload and the checksum of a Bech32 string.
Removing a length limit on the data part should therefore not include removing the length limit on the HRP.
Please correct me if I'm mistaken.

> "Note that Bech32 strings can only be all uppercase or all lowercase, but the checksum is always computed over the lowercase string."

I realise that and this is also what the Bech32 spec says (no mixed-case, lowercase for checksum computations), but the BIP also requires that the encoder never output an uppercase Bech32 string. Age does not follow this requirement (the output case follows the input case).
My understanding is that it does not matter in essence, but it is a deviation from the BIP and a comment (in code, not the spec) documenting this difference wouldn't hurt.

**Reply by BenWestgate** [reply]
FWIW, a Bech32 string could be valid without any letters, so the case of an encoding is unknown, that's why it's normalized to lowercase.

**Reply by FiloSottile** [reply]
Oh sorry, my bad, I didn't realize this was about HRP length specifically. We don't limit the length of a plugin, so I guess it's unlikely but possible someone out there has a HRP longer than 83 characters. Any reason not to modify the spec to say we don't limit HRP length either?

For the encode output in uppercase, I don't see the problem, the function docs say "If the HRP is uppercase, the output will be uppercase." already and we do the spec-compliant thing for the checksum, which makes our implementation interoperable. Not sure why the spec has opinions on the API, but switching to `strings.Upper(bech32.Encode())` everywhere just because the spec said so seems silly, especially since the bech32 package is internal, not a general purpose Bech32 implementation.

**Reply by mlemanowicz** [reply]
Re. HRP: Saying that age lifts the length limits on the HRP and the data parts would be clear enough. 
Re. case: There is no problem. Don't change anything in the implementation or the age spec. This point was merely about the observed difference from the Bech32 spec which puts this API requirement for whatever reason.

To be perfectly frank, I'm not thrilled with the quality of the Bech32 spec.
It is not precise and the reference implementations don't follow it to the letter.
The reference and Bitcoin Core C++ implementations don't even run lowercase on input in Encode, because they decide to just disallow uppercase characters (against the spec) in the input altogether.
